psamm/moma.py

```diff
@@ -79,84 +100,332 @@ def __init__(self, model, solver):
+        for f_reaction, f_value in iteritems(wt_fluxes):
+            # Makes sure the reaction we are trying to put a constraint on
+            # is a non exchange reaction
+            if f_reaction in reactions:
+                # Add the constraint that finds the optimal solution, such
+                # that the difference between the wildtype flux
+                # is similar to the knockout flux.
+                constr.add(
+                    z(f_reaction) >= v_wt(f_reaction) - v(f_reaction),
+                    v_wt(f_reaction) - v(f_reaction) >= -z(f_reaction))
+
+        # Set the objective for the sum of z
         self._prob.set_objective(z.sum(self._adjustment_reactions()))
 
+        # Set a constraint that says the biomass flux of the wild type is
+        # the same as the FBA analysis
         v_wt_obj = self._v_wt(objective)
         with self.constraints(v_wt_obj == wt_obj):
             result = self._solve(lp.ObjectiveSense.Minimize)
 
+        # Raise any errors
         if not result:
             raise MOMAError('Unable to solve L1 MOMA: {}'.format(
                 result.status))
 
+    def minimize_qlp2(self, objective, wt_fluxes):
+        """Implemetnation of the MOMA algorithm.
+
+        We try to maximize biomass, by keeping the fluxes relitively the same
+        as the wildtype. This helps avoid the assumption that an organism will
+        perform optimally directly after removing a gene.
+
+        Minimizes sum of (wild type - knockout)^2
+
+        Args:
+            objective: Biomass reaction for the model.
+            wt_fluxes: All the wt_fluxes found by FBA in the form of a
+                dictionary. Use :meth: _get_fba_flux(objective) to return
+                a dictionary.
+        """
+        # These are all of our non eachange reactions
+        reactions = set(self._adjustment_reactions())
+
+        # Get the v model and all the reactions 'variables' associated with it
+        v = self._v
+
+        # Create a constraints object for this model
+        constr = self.constraints()
+
+        obj_expr = 0
+        # Loop through all the flux reactions
+        for f_reaction, f_value in iteritems(wt_fluxes):
+            # Makes sure the reaction we are trying to put a constraint on is
+            # a non exchange reaction
+            if f_reaction in reactions:
+                # Add the constraint that finds the optimal solution, such
+                # that the differenct between the wildtype flux
+                # is similar to the knockout flux.
+                obj_expr += (f_value - v(f_reaction))**2
+
+        # Set the objective
+        self._prob.set_objective(obj_expr)
+
+        # Minimize the squared distance between the wild type flux values and
+        # the knockout values
+        result = self._solve(lp.ObjectiveSense.Minimize)
+
+        # We need to go back and manually delete all the constaints
+        constr.delete()
```

Comment:
Here the constraints are deleted before the `MOMAError` is potentially raised which is better and should in most cases produce the correct behavior when reusing the instance. Though, if an exception is thrown anywhere above we still get an inconsistent state. Use the `try`-block, 'finally'-block to handle this.